src/synth/fluid_voice.c

```diff
@@ -1161,25 +1175,32 @@ fluid_voice_update_param(fluid_voice_t *voice, int gen)
  * iteration of the audio cycle (which would probably be feasible if
  * the synth was made in silicon).
  *
- * The update is done in three steps:
+ * The update is done in four steps:
  *
- * - first, we look for all the modulators that have the changed
+ * - step 1: first, we look for all the modulators that have the changed
  * controller as a source. This will yield a list of generators that
  * will be changed because of the controller event.
  *
- * - For every changed generator, calculate its new value. This is the
- * sum of its original value plus the values of al the attached
+ * - step 2: For every changed generator, calculate its new value. This is the
+ * sum of its original value plus the values of all the attached
  * modulators.
  *
- * - For every changed generator, convert its value to the correct
- * unit of the corresponding DSP parameter
+ * - step 3: We need to avoid the risk to call 'fluid_voice_update_param' several
+ * times for the same generator if several modulators have that generator as
+ * destination. So every changed generators are registered in a list.
+
+ * - step 4: When finished with all the modulators, for every registered generator
+ * in the list, convert its value to the correct unit of the corresponding DSP
+ * parameter (fluid_voice_update_param()).
  */
 int fluid_voice_modulate(fluid_voice_t *voice, int cc, int ctrl)
 {
     int i, k;
     fluid_mod_t *mod;
     int gen;
     fluid_real_t modval;
+    int gen_changed[GEN_LAST]; /* registered list of changed generators */
```

Comment:
```suggestion
    enum fluid_gen_type gen_changed[GEN_LAST]; /* registered list of changed generators */
```

Consider to specify the enum type explicitly. The `enum` could be one byte wide, while the `int` is 4 or 8 bytes, in which case we would waste many bytes on the stack.